crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1093 @@
+        .filter(|(_, v)| is_supercontrolled(v))
+        .collect();
+
+    for basis_1q in &available_1q_basis {
+        for basis_2q in supercontrolled_basis.keys() {
+            let mut basis_2q_fidelity: f64 = match available_2q_props.get(basis_2q) {
+                Some(&(_, Some(e))) => 1.0 - e,
+                _ => 1.0,
+            };
+            if let Some(approx_degree) = approximation_degree {
+                basis_2q_fidelity *= approx_degree;
+            }
+            let gate = &supercontrolled_basis[basis_2q];
+            let decomposer = TwoQubitBasisDecomposer::new_inner(
+                gate.operation.name().to_owned(),
+                gate.operation.matrix(&gate.params).unwrap().view(),
+                basis_2q_fidelity,
+                basis_1q,
+                None,
+            )?;
+
+            decomposers.push(DecomposerElement {
+                decomposer: DecomposerType::TwoQubitBasisDecomposer(Box::new(decomposer)),
+                gate: gate.clone(),
+            });
+        }
+    }
+
+    // If our 2q basis gates are a subset of cx, ecr, or cz then we know TwoQubitBasisDecomposer
+    // is an ideal decomposition and there is no need to bother calculating the XX embodiments
+    // or try the XX decomposer
+    let available_basis_set: HashSet<&str> = available_2q_basis.keys().copied().collect();
+
+    #[inline]
+    fn check_goodbye(basis_set: &HashSet<&str>) -> bool {
+        !basis_set.iter().any(|gate| !GOODBYE_SET.contains(gate))
+    }
+
+    if check_goodbye(&available_basis_set) {
+        return Ok(Some(decomposers));
+    }
+
+    // Let's now look for possible controlled decomposers (i.e. XXDecomposer)
+    let controlled_basis: IndexMap<&str, NormalOperation> = available_2q_basis
+        .into_iter()
+        .filter(|(_, v)| is_controlled(v))
+        .collect();
+    let mut pi2_basis: Option<&str> = None;
+    let xx_embodiments: &Bound<'_, PyAny> = imports::XX_EMBODIMENTS.get_bound(py);
+
+    // The xx decomposer args are the interaction strength (f64), basis_2q_fidelity (f64),
+    // and embodiments (Bound<'_, PyAny>).
+    let xx_decomposer_args = controlled_basis.iter().map(
+        |(name, op)| -> PyResult<(f64, f64, pyo3::Bound<'_, pyo3::PyAny>)> {
+            let strength = 2.0
+                * TwoQubitWeylDecomposition::new_inner(
+                    op.operation.matrix(&op.params).unwrap().view(),
+                    None,
+                    None,
+                )
+                .unwrap()
+                .a();
+            let mut fidelity_value = match available_2q_props.get(name) {
+                Some(&(_, error)) => 1.0 - error.unwrap_or(0.0),
+                None => 1.0,
+            };
+            if let Some(approx_degree) = approximation_degree {
+                fidelity_value *= approx_degree;
+            }
+            let mut embodiment =
+                xx_embodiments.get_item(op.clone().into_py(py).getattr(py, "base_class")?)?;
+
```

Comment:
This capability of `NormalOperation` might be modified later. But you might prefer using `to_object(py)` here instead, since you'd use `clone_ref(py)` to perform the cloning of the `PyObject` in thi case.
```suggestion
                xx_embodiments.get_item(op.to_object(py).getattr(py, "base_class")?)?;
```